src/upgrade.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling Kafka in-place upgrades."""
+
+import logging
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+    verify_requirements,
+)
+from pydantic import BaseModel
+from typing_extensions import override
+
+from utils import get_zookeeper_version
+
+if TYPE_CHECKING:
+    from charm import KafkaCharm
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaDependencyModel(BaseModel):
+    """Model for Kafka Operator dependencies."""
+
+    service: DependencyModel
+
+
+class KafkaUpgrade(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "KafkaCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    @property
+    def idle(self) -> bool:
+        """Checks if cluster state is idle.
+
+        Returns:
+            True if cluster state is idle. Otherwise False
+        """
+        return self.cluster_state == "idle"
+
+    @property
+    def current_version(self) -> str:
+        """Get current Kafka version."""
+        dependency_model: DependencyModel = getattr(self.dependency_model, "service")
+        return dependency_model.version
+
+    @property
+    def zookeeper_current_version(self) -> str:
+        """Get current Zookeeper version."""
+        return get_zookeeper_version(zookeeper_config=self.charm.kafka_config.zookeeper_config)
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        default_message = "Pre-upgrade check failed and cannot safely upgrade"
+        if not self.charm.healthy:
+            raise ClusterNotReadyError(message=default_message, cause="Cluster is not healthy")
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        units = set([self.charm.unit] + list(self.charm.peer_relation.units))  # type: ignore[reportOptionalMemberAccess]
+        for unit in units:
+            upgrade_stack.append(int(unit.name.split("/")[-1]))
+
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        logger.warning("SOME USEFUL INSTRUCTIONS")  # TODO
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        dependency_model: DependencyModel = getattr(self.dependency_model, "service")
```

Comment:
**niptick** this lines looks like 52. This may call for a cached_property, also providing already the typing of the property in "DependencyModel" (which is always sort of forced)

Reply:
This is done so the linter doesn't go crazy. In theory, I can just call `self.dependency_model.service` but that line will say that `BaseModel` doesn't have `service`
In short, I don't really need to make an extra property out of it, because it already is a property of `DataUpgrade` class.